Chapter-4-applying-tactical-domain-driven-design/Fitnet.Common/Fitnet.Common.Core/ValueObject.cs

```diff
@@ -0,0 +1,26 @@
+﻿namespace EvolutionaryArchitecture.Fitnet.Common.Core;
+
+public abstract class ValueObject
+{
+    protected abstract IEnumerable<object> GetEqualityComponents();
+
+    public override bool Equals(object? obj)
+    {
+        if (obj == null || obj.GetType() != GetType())
+        {
+            return false;
+        }
+
+        var other = (ValueObject)obj;
+
+        return GetEqualityComponents().SequenceEqual(other.GetEqualityComponents());
+    }
+
+    public override int GetHashCode() => GetEqualityComponents()
+            .Aggregate(0, (current, obj) => current ^ (obj?.GetHashCode() ?? 0));
```

Comment:
public override int GetHashCode() =>
        GetAtomicValues().Aggregate(
            default(int),
            (hashcode, value) =>
                HashCode.Combine(hashcode, value.GetHashCode()));
                
                We could use potentially HashCode.Combine to make this caculation simpler. What you think?